### PR TITLE
Merge test fixes

### PR DESCRIPTION
**Pull request by abergeron**
These are most of the test fixes that I will need for multi-gpu for code that is outside sandbox/gpuarray/

### CODE REVIEW

**Comment by abergeron** [comment]
This would need a test run on a GPU to make sure that stuff in sandbox/cuda and sandbox/gpuarray is not broken.

File: theano/sandbox/cuda/tests/test_basic_ops.py [hunk 1]
```diff
@@ -945,14 +945,14 @@ class TestAlloc(theano.tensor.tests.test_basic.TestAlloc):
     dtype = "float32"
     mode = mode_with_gpu
     shared = staticmethod(cuda.shared_constructor)
-    allocs = [B.GpuAlloc, B.GpuAlloc, tensor.Alloc]
+    allocs = [B.GpuAlloc(), B.GpuAlloc(), tensor.Alloc()]
```
Comment (lamblin) [inline theano/sandbox/cuda/tests/test_basic_ops.py]: Why is that needed? Just to avoid calling the constructor too many times?

File: theano/sandbox/gpuarray/tests/test_basic_ops.py [hunk 2]
```diff
@@ -304,7 +305,7 @@ class TestAlloc(theano.tensor.tests.test_basic.TestAlloc):
     dtype = "float32"
     mode = mode_with_gpu
     shared = staticmethod(gpuarray_shared_constructor)
-    allocs = [GpuAlloc, GpuAlloc, T.Alloc]
+    allocs = [GpuAlloc(), GpuAlloc(), T.Alloc]
```
Comment (lamblin) [inline theano/sandbox/gpuarray/tests/test_basic_ops.py]: Why not `T.Alloc()`?

File: theano/sandbox/gpuarray/tests/test_basic_ops.py [hunk 3]
```diff
@@ -340,30 +341,29 @@ def test_gpu_contiguous():
...
-class G_Join_and_Split(T_Join_and_Split):
+class G_Join_and_Split(test_basic.T_Join_and_Split):
     def setUp(self):
         super(G_Join_and_Split, self).setUp()
         self.mode = mode_with_gpu.excluding('constant_folding')
-        self.join_op = GpuJoin
+        self.join_op = GpuJoin()
         self.split_op = GpuSplit
         # Use join instead of MakeVector since there is no MakeVector on GPU
         self.make_vector_op = GpuJoin
```
Comment (lamblin) [inline theano/sandbox/gpuarray/tests/test_basic_ops.py]: Why not `GpuJoin()` here?

Comment (lamblin) [inline theano/sandbox/gpuarray/tests/test_basic_ops.py]: It is kind of confusing to have some Op classes and some Op instances, all named `..._op`.
I understand that we cannot pre-instantiate `GpuSplit` because the number of splits is a constructor argument, but we should name the members differently then.

**Comment by lamblin** [comment]
Otherwise, it looks OK, as long as it passes on a machine with a GPU and gpuarray.

File: theano/sandbox/cuda/tests/test_basic_ops.py [hunk 4]
```diff
@@ -945,14 +945,14 @@ class TestAlloc(theano.tensor.tests.test_basic.TestAlloc):
...
 
 
 class T_Join_and_Split(theano.tensor.tests.test_basic.T_Join_and_Split):
     def setUp(self):
         utt.seed_rng()
         self.mode = mode_with_gpu.excluding('constant_folding')
-        self.join_op = cuda.GpuJoin
+        self.join_op = cuda.GpuJoin()
         # No gpu split.
         self.split_op = tensor.Split
```
Comment (nouiz) [inline theano/sandbox/cuda/tests/test_basic_ops.py]: This line is a good example of possible confusion @lamblin talk about. Can we at least make it consistent inside one class? I'm ok to not change the name if it is consistent in one class. Do you know if other class still use the old definition? If yes, we could make a ticket to make this constant.

**Comment by nouiz** [comment]
After the small change, I'll run all tests (CPU, old GPU, new GPU back-end)

File: theano/sandbox/cuda/tests/test_basic_ops.py (hunk 1 above)
Comment (abergeron) [inline theano/sandbox/cuda/tests/test_basic_ops.py]: Because the future GpuAlloc op will require some parameters to create (namely the context).  It isn't convenient to pass that to the test, so it's simple to have instances than classes.

File: theano/sandbox/gpuarray/tests/test_basic_ops.py (hunk 2 above)
Comment (abergeron) [inline theano/sandbox/gpuarray/tests/test_basic_ops.py]: Because I forgot, I'll fix it.

File: theano/sandbox/gpuarray/tests/test_basic_ops.py (hunk 3 above)
Comment (abergeron) [inline theano/sandbox/gpuarray/tests/test_basic_ops.py]: Yes, you're right.  make_vector_op should be an instance too.

As for split_op, we can rename it, I just need a good suggestion.

**Comment by abergeron** [comment]
I've made most of the fixes, except the rename

File: theano/sandbox/gpuarray/tests/test_basic_ops.py (hunk 3 above)
Comment (lamblin) [inline theano/sandbox/gpuarray/tests/test_basic_ops.py]: `split_op_class`? `split_op_type`?

**Comment by lamblin** [comment]
For convenience:
```
======================================================================
FAIL: test_stack_mixed_type_constants (theano.tensor.tests.test_basic.T_Join_and_Split)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/miniconda/envs/pyenv/lib/python2.6/site-packages/theano/tensor/test/test_basic.py", line 3243, in test_stack_mixed_type_constants
    out = self.eval_outputs_and_check_vector([s], opt.MakeVector)
  File "/home/travis/miniconda/envs/pyenv/lib/python2.6/site-packages/theano/tensor/test/test_basic.py", line 3221, in eval_outputs_and_check_vector
    if isinstance(node.op, type(make_vector_op))]
AssertionError
```

**Comment by abergeron** [comment]
I rebased the PR, fixed the error with make_vector and renamed split_op to split_op_class.

Now it should be ready (I hope).
